sqlglot/expressions.py

```diff
@@ -3555,7 +3589,32 @@ def build(
         return DataType(**{**data_type_exp.args, **kwargs})
 
     def is_type(self, *dtypes: str | DataType | DataType.Type) -> bool:
-        return any(self.this == DataType.build(dtype).this for dtype in dtypes)
+        """
+        Checks whether this DataType matches one of the provided data types. Nested types like
+        arrays or structs will be compared using "structural equivalence" semantics, so e.g.
+        array<int> != array<float>.
+
+        Args:
+            dtypes: the data types to compare this DataType to.
+
+        Returns:
+            True, if and only if there is a type in `dtypes` which is equal to this DataType.
+        """
+        if self.expressions and self.this in DataType.NESTED_TYPES:
```

Comment:
why the self.this in datatype.nested_types?

doesn't that mean

float(3.0) == float(12.0)

Reply:
I see, I misinterpreted the above suggestion - will update